### PR TITLE
Ch14413/vad for ac

### PR DESCRIPTION
**Pull request by karthikbr82**

### CODE REVIEW

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 1]
```diff
@@ -1,6 +1,92 @@
...
+      this.analyser.smoothingTimeConstant = 0.1;
+      this.fftBins = new Float32Array(this.analyser.frequencyBinCount);
+      this.mediaStreamSource = this.audioContext.createMediaStreamSource(stream);
+      this.mediaStreamSource.connect(this.analyser);
+      this.isSpeaking = false;
+      this.maxVolumeHistory = [];
+      for (var i=0; i < 10; i++) {
+        this.maxVolumeHistory.push(0);
+      }
+      this.threshold = -50;
```
Comment (marcinag) [inline amazon-connect/callstats-amazon-connect-shim.js:19]: Can we add comments about these parameters here?

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 2]
```diff
@@ -1,6 +1,92 @@
...
+      this.callback = callback;
+      this.start();
+    }
+
+    getMaxVolume () {
+      var maxVolume = -Infinity;
+      this.analyser.getFloatFrequencyData(this.fftBins);
+    
+      for(var i=4, ii=this.fftBins.length; i < ii; i++) {
+        if (this.fftBins[i] > maxVolume && this.fftBins[i] < 0) {
```
Comment (marcinag) [inline amazon-connect/callstats-amazon-connect-shim.js:30]: I don't understand where does this.fftBins[i] < 0 come from? Why does it have to be negative?

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 3]
```diff
@@ -1,6 +1,92 @@
...
+    }
+
+    detect() {
+      var maxVolume = this.getMaxVolume();
+      var totalMaxVolume = 0;
+      if (maxVolume > this.threshold && !this.speaking) {
+        for (var i = this.maxVolumeHistory.length - 3; i < this.maxVolumeHistory.length; i++) {
+          totalMaxVolume += this.maxVolumeHistory[i];
+        }
+        if (totalMaxVolume >= 2) {
```
Comment (marcinag) [inline amazon-connect/callstats-amazon-connect-shim.js:54]: Where does 2 come from?

Comment (marcinag) [reply amazon-connect/callstats-amazon-connect-shim.js:54]: So we're pushing only 1s to the maxVolumeHistory array. Perhaps we rename it to something else like volumeHistory, as there's no max in this data in my opinion.

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 4]
```diff
@@ -136,6 +265,59 @@
       }
     }
 
+    function sendCustomEvent(eventType) {
+      if (prevSpeakingState === eventType) {
+        return;
+      }
+      console.log(eventType);
```
Comment (marcinag) [inline amazon-connect/callstats-amazon-connect-shim.js]: We can probably remove it

File: amazon-connect/callstats-amazon-connect-shim.js (hunk 1 above)
Comment (marcinag) [reply amazon-connect/callstats-amazon-connect-shim.js:19]: Where is the -50 coming from? Are all values from FFT negative here?

Comment (marcinag) [reply amazon-connect/callstats-amazon-connect-shim.js:19]: Conversation is considered to be between 40-60dB, so I wonder if it 50 and results of FFT are negative?

**Review by marcinag — APPROVED** [review]
LGTM. But I left some comments.

File: amazon-connect/callstats-amazon-connect-shim.js (hunk 1 above)
Comment (karthikbr82) [reply amazon-connect/callstats-amazon-connect-shim.js:19]: webAudio API defines AnalyserNode,

It has the following attributes
maxDecibels, of type double
maxDecibels is the maximum power value in the scaling range for the FFT analysis data for conversion to unsigned byte values. The default value is -30. If the value of this attribute is set to a value less than or equal to minDecibels, an IndexSizeError exception MUST be thrown.
minDecibels, of type double
minDecibels is the minimum power value in the scaling range for the FFT analysis data for conversion to unsigned byte values. The default value is -100. If the value of this attribute is set to a value more than or equal to maxDecibels, an IndexSizeError exception MUST be thrown.

So the default range is -100 to -30
0 dB is the loudest possible sound, -10 dB is a 10th of that

-50 is the threshold if the speech is present in the stream.

File: amazon-connect/callstats-amazon-connect-shim.js (hunk 2 above)
Comment (karthikbr82) [reply amazon-connect/callstats-amazon-connect-shim.js:30]: getFloatFrequencyData(array) gives us the fftbins which are of the range minDecibels - maxDecibels which is -100 to -30

**Review by marcinag — APPROVED** [review]
LGTM

**Pull request merged**